synapse/lib/snap.py

```diff
@@ -1100,94 +1199,51 @@ async def _addNodeDef(self, nodedefn):
+                for tagname, tagvalu in tags.items():
+                    await protonode.addTag(tagname, tagvalu)
 
-        nodedata = forminfo.get('nodedata')
-        if nodedata is not None:
-            try:
-                for name, data in nodedata.items():
-                    # make sure we have valid nodedata
-                    if not (isinstance(name, str)):
-                        await self.warn(f'Nodedata key is not a string: {name}')
+            nodedata = forminfo.get('nodedata')
+            if isinstance(nodedata, dict):
+                for dataname, datavalu in nodedata.items():
+                    if not isinstance(dataname, str):
                         continue
-                    s_common.reqjsonsafe(data)
-                    edits.append((s_layer.EDIT_NODEDATA_SET, (name, data, None), ()))
-            except asyncio.CancelledError:  # pragma: no cover  TODO:  remove once >= py 3.8 only
-                raise
-            except Exception as e:
-                await self.warn(f'Failed adding node data on {formname}, {formvalu}, {forminfo}: {e}')
-
-        n2edits = []
-        for verb, n2iden in forminfo.get('edges', ()):
-            # check for embedded ndef rather than n2iden
-            if isinstance(n2iden, (list, tuple)):
-                n2formname, n2valu = n2iden
-
-                n2form = self.core.model.form(n2formname)
-                if n2form is None:
-                    await self.warn(f'Failed to make n2 edge node for {n2iden}: invalid form')
-                    continue
+                    await protonode.setData(dataname, datavalu)
 
-                if n2form.isrunt:
-                    await self.warn(f'Edges cannot be used with runt nodes: {n2formname}')
-                    continue
+            tagprops = forminfo.get('tagprops')
+            if tagprops is not None:
+                for tag, props in tagprops.items():
+                    for name, valu in props.items():
+                        await protonode.setTagProp(tag, name, valu)
 
-                try:
-                    n2norm, _ = n2form.type.norm(n2valu)
-                    n2buid = s_common.buid((n2form.name, n2norm))
+            for verb, n2iden in forminfo.get('edges', ()):
 
-                    if n2buid not in n2buids:
-                        _, n2editset = await self._getAddNodeEdits(n2formname, n2valu)
-                        n2edits.extend(n2editset)
+                if isinstance(n2iden, (tuple, list)):
+                    n2proto = await editor.addNode(*n2iden)
+                    if n2proto is None:
+                        continue
 
-                except asyncio.CancelledError:  # pragma: no cover  TODO:  remove once >= py 3.8 only
-                    raise
+                    n2iden = n2proto.iden()
 
-                except:  # pragma: no cover
-                    await self.warn(f'Failed to make n2 edge node for {n2iden}')
+                if not isinstance(verb, str):
```

Comment:
should we check that verb is valid before adding the n2 node?

Reply:
I dont think we currently restrict the strings that can be used for edge verbs ( at least with vars ) right?